api/tests/announcement_test.py

```diff
@@ -0,0 +1,228 @@
+""" Test for announcement APIs """
+import requests
+
+from tests import SERVER, ADMIN_ACCESS, BASE_ACCESS
+from tests.framework import assert_status_code
+
+def add_announcement():
+    """ Add a new announcement """
+    response = requests.get(SERVER+"/api/department/1/announcements",
+                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
+                   data={})
+    assert_status_code(response, 200)
+    data = response.json()
+    initial_ids = []
+    for item in data['data']:
+        initial_ids.append(item['id'])
+
+    response = requests.post(SERVER+"/api/announcement/1",
+                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
+                   data={'Scope':'2',
+                         'Text': 'testing',
+                         'Email': '0'})
+    assert_status_code(response, 201)
+
+    response = requests.get(SERVER+"/api/department/1/announcements",
+                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
+                   data={})
+    assert_status_code(response, 200)
+    data = response.json()
+    # Find New Index
+    target = None
+    for item in data['data']:
+        if target is None and item['id'] not in initial_ids:
+            target = item['id']
+
+    response = requests.get(SERVER+"/api/announcement/"+target,
+                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
+                   data={})
+    assert_status_code(response, 200)
+    item = response.json()
+    assert 'type' in item
+    assert item['type'] == 'announcement'
+    assert 'id' in item
+    assert item['id'] == target
+    assert 'departmentId' in item
+    assert item['departmentId'] == '1'
+    assert 'postedOn' in item
+    assert 'postedBy' in item
+    assert 'scope' in item
+    assert item['scope'] == '2'
+    assert 'text' in item
+    assert item['text'] == 'testing'
+
+    return target
+
+def test_announce_priv():
+    """ Getting the announcments as a privilaged user """
+
+    target = add_announcement()
+
+    response = requests.get(SERVER+"/api/department/1/announcements",
+                   headers={'Authorization': 'Bearer '+ADMIN_ACCESS},
+                   data={})
+    assert_status_code(response, 200)
+    data = response.json()
+    assert 'data' in data
```

Comment:
The more usual way to test that a response body has been fulfilled is to map them both to hashes/dictionaries and compare those, using whatever comparison operations the language provides.

So, you'd say something like:
```python
expected_foo = {
  bar: 'baz',
  quux: 'quuux',
  zorch: {
    spaz: 'squid'
  }
}

response = requests.get(...)
data = response.json()

# whatever you do to compare hashes in Python goes here
```

Reply:
I agree the long list of asserts sucks and a better way to frame it would be welcome.

Some web searching says that this can be messy, because json can be reordered, and if I am reading this correctly we could get into trouble with that (https://stackoverflow.com/questions/25851183/how-to-compare-two-json-objects-with-the-same-elements-in-a-different-order-equa)

Secondly there are fields we do not care about and do not want to compare, date fields spring to mind. So we would have to handle those fields.   It just makes the == a much harder operation than would be hoped. 

Unless you know magic to fix that. :)

Reply:
AFAIK, python set operators don't care about order, and dictionary comparisons of keys and values are treated as set comparisons.

```python
>>> {'bar': 'foo', 'baz': 'quux'} == {'baz': 'quux', 'bar': 'foo'}                                                      
True 
```

The same is true if you just compare keys:

```python
>>> {'bar': 'foo', 'baz': 'quux'}.keys() == {'baz': 'quux', 'bar': 'foo'}.keys()
True 
```

So the real question is, what are we testing here. If we're testing structure, we can use `keys()` to do so and still get down to not having to line-by-line compare. If it's values we care about, we should be unit testing the functions that return the values. If we simply can't do that, then we should be testing values we've inserted into the database ourselves, as the 'before' part of a test, which means we can in fact compare the dates with accuracy, because they'll be dates we just inserted explicitly.